open-build-service-mock/src/api/mod.rs

```diff
@@ -0,0 +1,115 @@
+use std::fmt::Display;
+
+use http_types::{auth::BasicAuth, StatusCode};
+use wiremock::{Request, ResponseTemplate};
+use xml_builder::XMLElement;
+
+mod build;
+pub(crate) use build::*;
+
+mod source;
+pub(crate) use source::*;
+
+trait ResponseTemplateUtils {
+    fn set_body_xml(self, xml: XMLElement) -> Self;
+}
+
+impl ResponseTemplateUtils for ResponseTemplate {
+    fn set_body_xml(self, xml: XMLElement) -> Self {
+        let mut body = vec![];
+        xml.render(&mut body, false, true).unwrap();
+        self.set_body_raw(body, "application/xml")
+    }
+}
+
+#[derive(Debug)]
+struct ApiError {
+    http_status: StatusCode,
+    code: String,
+    summary: String,
+}
+
+impl ApiError {
+    fn new(http_status: StatusCode, code: String, summary: String) -> ApiError {
+        ApiError {
+            http_status,
+            code,
+            summary,
+        }
+    }
+
+    fn into_response(self) -> ResponseTemplate {
+        let mut status_xml = XMLElement::new("status");
+        status_xml.add_attribute("code", &self.code);
+
+        let mut summary_xml = XMLElement::new("summary");
+        summary_xml.add_text(self.summary).unwrap();
+
+        status_xml.add_child(summary_xml).unwrap();
+        ResponseTemplate::new(self.http_status).set_body_xml(status_xml)
+    }
+}
+
+impl Display for ApiError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}: {}: {}", self.http_status, self.code, self.summary)
+    }
+}
+
+fn unknown_project(project: String) -> ApiError {
```

Comment:
Originally these were indeed on `ApiError`, but I ended up making it top-level the two files that are endpoint-specific have some of their own functions to return errors, too, so that would end up with one of these scenarios:

- some functions on `ApiError` and some function top-level
- adding "dummy" traits to extend `ApiError` in the individual files
- move all the error functions to `mod.rs`

though tbh I'm partial to the last one in particular, in the end it mostly boiled down to the flip of a coin. Would moving them to `mod.rs` so that they can all be under `ApiError` be good here?

Reply:
I guess i would veer towards ApiError having the generic error constructors in the impl block and error that are really speciifc to specific api's as helper functions; but this can be done later.